### PR TITLE
Fixed IAAS-10652: Added check to verify database existence

### PR DESCRIPTION
**Pull request by shahpratikr**

### CODE REVIEW

File: cmd/createdb.go [hunk 1]
```diff
@@ -28,9 +27,14 @@ var createDBCmd = &cobra.Command{
         if userName == "" {
             userName = serviceName
         }
-        err := cfgMgr.CreateDB(serviceName, userName)
+        dbPassword := cfgMgr.GetRandomPassword()
+        dbCreated, err := cfgMgr.CreateDB(serviceName, userName)
```
Comment (pacharya-pf9) [inline cmd/createdb.go:31]: This error is not being checked

File: pkg/cfg/cfgmgr.go [hunk 2]
```diff
@@ -163,71 +165,105 @@ func (c * CfgMgr) getRandomPassword() string {
...
-func (c *CfgMgr) CreateDB(serviceName, userName string) error {
+func (c *CfgMgr) CreateDB(serviceName, userName string) (bool, error) {
+    dbObject, err := c.getDbObject()
+    if err != nil {
+        return false, err
+    }
+    _, err = dbObject.Exec(fmt.Sprintf("SHOW DATABASE LIKE %s", serviceName))
+    if err == nil {
+        zap.L().Info(fmt.Sprintf("Database %s already exists", serviceName))
+        return false, nil
```
Comment (pacharya-pf9) [inline pkg/cfg/cfgmgr.go:176]: Isn't DB technically created at this point? Can you return `true, nil` to that effect?

File: pkg/cfg/cfgmgr.go [hunk 3]
```diff
@@ -163,71 +165,105 @@ func (c * CfgMgr) getRandomPassword() string {
     return string(out)
 }
 
-func (c *CfgMgr) CreateDB(serviceName, userName string) error {
+func (c *CfgMgr) CreateDB(serviceName, userName string) (bool, error) {
```
Comment (pacharya-pf9) [inline pkg/cfg/cfgmgr.go]:
```suggestion
func (c *CfgMgr) CreateDB(serviceName, userName string) (updateConsul bool, err error) {
```
Anyone using the function will have better understanding of what the return types represent

File: pkg/cfg/cfgmgr.go [hunk 4]
```diff
@@ -163,71 +165,105 @@ func (c * CfgMgr) getRandomPassword() string {
...
+    _, err = dbObject.Exec(fmt.Sprintf("CREATE DATABASE %s", serviceName))
+    if err != nil {
+        zap.L().Error("Error while creating database", zap.Error(err))
+        return false, err
+    }
+    zap.L().Info(fmt.Sprintf("Created DB '%s' successfully", serviceName))
+    return true, nil
+}
+
+func (c *CfgMgr) CreateGrants(serviceName, userName, dbPassword string) (bool, error) {
```
Comment (pacharya-pf9) [inline pkg/cfg/cfgmgr.go]:
```suggestion
func (c *CfgMgr) CreateGrants(dbName, userName, dbPassword string) (bool, error) {
```

File: pkg/cfg/cfgmgr.go [hunk 5]
```diff
@@ -163,71 +165,105 @@ func (c * CfgMgr) getRandomPassword() string {
...
+    updated := false
+    for rows.Next() {
+        _ = rows.Scan(&hostname)
+    }
+    hosts := []string{"localhost", "%", hostname}
+    for _, hostName := range hosts {
+        before_grants := c.getGrants(userName, hostName, dbObject)
+        query := fmt.Sprintf("GRANT ALL PRIVILEGES ON %s.* TO '%s'@'%s' IDENTIFIED BY '%s'",
+                             serviceName, userName, hostName, dbPassword)
+        _, _ = dbObject.Exec(query)
```
Comment (pacharya-pf9) [inline pkg/cfg/cfgmgr.go:209]: Are you sure that the results, output and error, of this query can be ignored completely?

File: pkg/cfg/cfgmgr.go [hunk 6]
```diff
@@ -163,71 +165,105 @@ func (c * CfgMgr) getRandomPassword() string {
...
+        after_grants := c.getGrants(userName, hostName, dbObject)
+        if (reflect.DeepEqual(before_grants, after_grants)) {
+            updated = false
+        } else {
+            updated = true
+        }
+    }
+    return updated, nil
```
Comment (pacharya-pf9) [inline pkg/cfg/cfgmgr.go]: If the last grant fails and all others are successful then this function will return `false`. Similarly if the grant fails for the first and second but succeeds for the last one, this function will return `true` incorrectly.

File: pkg/cfg/cfgmgr.go [hunk 7]
```diff
@@ -238,8 +274,25 @@ func (c *CfgMgr) CreateDB(serviceName, userName string) error {
     }
     _, _, _, err = c.ConsulKV.Txn(ops, nil)
     if err != nil {
-        zap.L().Error("Can't write db config to Consul, please retry later")
+        zap.L().Error("Can't write db config to Consul", zap.Error(err))
```
Comment (pacharya-pf9) [inline pkg/cfg/cfgmgr.go:279]:
```suggestion
        zap.L().Error("Can't write db config to Consul", err)
```
Does this not work?

File: pkg/cfg/cfgmgr.go [hunk 8]
```diff
@@ -238,8 +274,25 @@ func (c *CfgMgr) CreateDB(serviceName, userName string) error {
...
         return err
     }
     return nil
 }
+
+func (c *CfgMgr) getGrants(userName, host string, dbObject *sql.DB) []string {
+    var grants[] string
```
Comment (pacharya-pf9) [inline pkg/cfg/cfgmgr.go]:
```suggestion
    var grants []string
```

File: pkg/cfg/cfgmgr.go (hunk 2 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:176]: false is returned as we have nothing to update for DB in consul.

File: pkg/cfg/cfgmgr.go (hunk 5 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:209]: yes. at this point, we have DB and user in place. hence it's okay to ignore error here.

File: pkg/cfg/cfgmgr.go (hunk 7 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:279]: no. it gave `pkg/cfg/cfgmgr.go:279:23: cannot use err (type error) as type zapcore.Field in argument to zap.L().Error` this error while compiling.

**Review by pacharya-pf9 — APPROVED** [review]
/lgtm
/approve
I am not so familiar with go-firkinize but the change itself looks fine.

**Pull request merged**